### PR TITLE
reland #12332: refactor(ext/http): rewrite hyper integration and fix bug

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

File: ext/http/01_http.js [hunk 1]
```diff
@@ -84,6 +84,11 @@
         throw error;
       }
       if (nextRequest == null) {
+        // Work-around for servers (deno_std/http in particular) that call
```
Comment (piscisaureus) [inline ext/http/01_http.js:87]: I believe this is indicative of a problem with our API, in particular two dubious practices clash here:
* `HttpConn` is automatically closed when `HttpConn.nextRequest()` fails or a return an error.
* `Deno.upgradeWebSocket()` doesn't actually do anything until the caller uses (returns) the response created by `Deno.upgradeWebSocket()`.

**Review by bartlomieju — APPROVED** [review]
LGTM

File: cli/tests/unit/http_test.ts [hunk 2]
```diff
@@ -1000,6 +1002,80 @@ unitTest(
...
+
+    await Promise.all([server(), delay(100).then(client)]);
+  },
+);
+
+// https://github.com/denoland/deno/pull/12704
+// https://github.com/denoland/deno/pull/12732
+unitTest(
+  { permissions: { net: true } },
+  async function httpConnAutoCloseDelayedOnUpgrade() {
```
Comment (piscisaureus) [inline cli/tests/unit/http_test.ts:1045]: Regression test for the deno_std/http problem that got denoland#12332 reverted.

File: ext/http/01_http.js (hunk 1 above)
Comment (lucacasonato) [reply ext/http/01_http.js:87]: > Deno.upgradeWebSocket() doesn't actually do anything until the caller uses (returns) the response created by Deno.upgradeWebSocket()

Why is this a problem? This was done intentionally. The `Deno.upgradeWebSocket` call is just a glorified `new Response()`.

**Review by lucacasonato — APPROVED** [review]

**Pull request merged**